internal/settings/providers/namecom/provider.go

```diff
@@ -0,0 +1,123 @@
+package namecom
+
+import (
+	"context"
+	"encoding/json"
+
+	stderrors "errors"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/settings/constants"
+	"github.com/qdm12/ddns-updater/internal/settings/errors"
+	"github.com/qdm12/ddns-updater/internal/settings/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+)
+
+type Provider struct {
+	domain    string
+	host      string
+	ipVersion ipversion.IPVersion
+	username  string
+	password  string
+	ttl       int
+}
+
+func New(data json.RawMessage, domain, host string,
+	ipVersion ipversion.IPVersion) (p *Provider, err error) {
+	extraSettings := struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+		TTL      int    `json:"ttl"`
+	}{}
+	err = json.Unmarshal(data, &extraSettings)
+	if err != nil {
+		return nil, err
+	}
+	p = &Provider{
+		domain:    domain,
+		host:      host,
+		ipVersion: ipVersion,
+		username:  extraSettings.Username,
+		password:  extraSettings.Password,
+		ttl:       extraSettings.TTL,
+	}
+	err = p.isValid()
+	if err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *Provider) isValid() error {
+	switch {
+	case p.username == "":
+		return fmt.Errorf("%w", errors.ErrEmptyUsername)
+	case p.password == "":
+		return fmt.Errorf("%w", errors.ErrEmptyPassword)
+	}
+	return nil
+}
+
+func (p *Provider) String() string {
+	return utils.ToString(p.domain, p.host, constants.NameCom, p.ipVersion)
+}
+
+func (p *Provider) Domain() string {
+	return p.domain
+}
+
+func (p *Provider) Host() string {
+	return p.host
+}
+
+func (p *Provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.host, p.domain)
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName())),
+		Host:      models.HTML(p.Host()),
+		Provider:  "<a href=\"https://name.com\">Name.com</a>",
+		IPVersion: models.HTML(p.ipVersion.String()),
+	}
+}
+
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip net.IP) (newIP net.IP, err error) {
+	// Documentation at https://www.name.com/api-docs
+	recordType := constants.A
+	if ip.To4() == nil {
+		recordType = constants.AAAA
+	}
+
+	recordID, err := p.getRecordID(ctx, client, recordType)
+
+	if err == nil && recordID > 0 {
+		err = p.updateRecord(ctx, client, recordID, ip)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", errors.ErrUpdateRecord, err)
+		}
+	}
+
+	if stderrors.Is(err, errors.ErrRecordNotFound) {
+		err = p.createRecord(ctx, client, ip)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", errors.ErrCreateRecord, err)
+		}
+	} else if err != nil {
+		return nil, fmt.Errorf("%w: %w", errors.ErrGetRecordID, err)
+	}
```

Comment:
nit it would be to flip this around and handle the error first and return earlier (i.e `return ip, nil`) so we have:
- less code blocks/indentation
- no need to check for `recordID > 0`
- Go standard programming approach is to handle errors first

```suggestion
if stderrors.Is(err, errors.ErrRecordNotFound) {
		err = p.createRecord(ctx, client, ip)
		if err != nil {
			return nil, fmt.Errorf("%w: %w", errors.ErrCreateRecord, err)
		}
		return ip, nil
	} else if err != nil {
		return nil, fmt.Errorf("%w: %w", errors.ErrGetRecordID, err)
	}

	err = p.updateRecord(ctx, client, recordID, ip)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", errors.ErrUpdateRecord, err)
	}
```